web/src/main/java/se/inera/intyg/webcert/web/auth/WebcertLoggingSessionRegistryImpl.java

```diff
@@ -31,53 +32,69 @@
-
         LOGGER.debug("Attempting to register new session '{}'", sessionId);
 
-        if (principal != null && principal instanceof WebCertUser) {
-            WebCertUser user = (WebCertUser) principal;
-            String userRole =
-                user.getRoles() != null && user.getRoles().size() == 1 ? user.getRoles().keySet().iterator().next() : "noRole?";
-            monitoringService
-                .logUserLogin(user.getHsaId(), userRole, user.getRoleTypeName(), user.getAuthenticationScheme(), user.getOrigin());
+        if (!isWebcertUser(principal)) {
+            return;
         }
 
+        final var user = (WebCertUser) principal;
+        final var userRole = getUserRole(user);
+        monitoringService.logUserLogin(user.getHsaId(), userRole, user.getRoleTypeName(), user.getAuthenticationScheme(), user.getOrigin());
+
         super.registerNewSession(sessionId, principal);
     }
 
     @Override
     public void removeSessionInformation(String sessionId) {
-
         LOGGER.debug("Attempting to remove session '{}'", sessionId);
 
-        SessionInformation sessionInformation = getSessionInformation(sessionId);
-
-        if (sessionInformation == null) {
+        final var session = sessionRepository.findById(sessionId);
+        final var user = getUser(session);
+        if (user == null) {
             super.removeSessionInformation(sessionId);
             return;
         }
 
-        Object principal = sessionInformation.getPrincipal();
-
-        if (principal instanceof WebCertUser) {
-            WebCertUser user = (WebCertUser) principal;
-            if (sessionInformation.isExpired()) {
-                monitoringService.logUserSessionExpired(user.getHsaId(), user.getAuthenticationScheme());
-            } else {
-                monitoringService.logUserLogout(user.getHsaId(), user.getAuthenticationScheme());
-            }
+        if (session.isExpired()) {
+            monitoringService.logUserSessionExpired(user.getHsaId(), user.getAuthenticationScheme());
+        } else {
+            monitoringService.logUserLogout(user.getHsaId(), user.getAuthenticationScheme());
         }
 
         super.removeSessionInformation(sessionId);
     }
 
+    private WebCertUser getUser(Session session) {
+        if (session == null) {
+            return null;
+        }
+        final var authenticator = (SecurityContext) session.getAttribute("SPRING_SECURITY_CONTEXT");
+        final var principal = authenticator != null ? authenticator.getAuthentication().getPrincipal() : null;
+        return isWebcertUser(principal) ? (WebCertUser) principal : null;
+    }
+
+    private boolean isWebcertUser(Object principal) {
+        return principal instanceof WebCertUser;
+    }
+
+    private static String getUserRole(WebCertUser user) {
+        return user.getRoles() != null && user.getRoles().size() == 1 ? user.getRoles().keySet().iterator().next() : "noRole?";
```

Comment:
Shouldn't it be size >= 1?

Reply:
It is logic that isn't touched in this PR (other than some refactoring to a private-method). But it seems like it is implemented as intended. If the user has "one" user-role, then it should be part of the loggpost. 

BaseUserDetailsService.decorateIntygUserWithRoleAndAuthorities sets the role, which is always only one.